### PR TITLE
Week6

### PR DESCRIPTION
**Pull request by andrii-zubenko**
- Added a custom topbar to `CountryInfoScreen` and `CountryDetailsScreen` where:
taps: is a StateFlow<Int>
backs: is a StateFlow<Int>
Unique countries selected: is a SharedFlow that saves only uniques countries that were selected from CountryInfoScreen
refresh:  button that triggers the loading screen and fetch from the API again
go dark/go light: button that uses StateFlow to change its own theme when tapped
Screenshot 2023-10-08 at 11 13 53 PM

- An app uptime counter on loading screen while data is being fetched from API
Screenshot 2023-10-08 at 11 11 59 PM

-created private method `getCountryInfoFlow` that is making an api call an return a Flow

### CODE REVIEW

File: app/src/main/java/com/kodeco/android/coordplot/MainActivity.kt [hunk 1]
```diff
@@ -26,6 +32,11 @@ class MainActivity : ComponentActivity() {
             CoordPlotTheme {
                 val configuration = LocalConfiguration.current
                 val navController = rememberNavController()
+                val tapState = tapFlow.collectAsState()
+                val backState = backFlow.collectAsState()
+                val countriesSelected =
+                    uniqueCountriesSelectedFlow.collectAsState(initial = emptyList())
+
                 NavHost(navController = navController, startDestination = "mainscreen") {
```
Comment (zlippard) [inline app/src/main/java/com/kodeco/android/coordplot/MainActivity.kt]: Great use of the NavHost already! Not a direct requirement here, but you could move this NavHost in to it's own `@Composable` function for better readability.

File: app/src/main/java/com/kodeco/android/coordplot/MainActivity.kt [hunk 2]
```diff
@@ -37,7 +48,20 @@ class MainActivity : ComponentActivity() {
                         PlotSurface(configuration.orientation)
                     }
                     composable("country_list") {
-                        CountryInfoScreen(navController)
+                        CountryInfoScreen(
+                            navigation = navController,
+                            countersTopBar = {
+                                CountersTopBar(
```
Comment (zlippard) [inline app/src/main/java/com/kodeco/android/coordplot/MainActivity.kt]: Nice work creating this CountersTapBar and reusing it on both the list and details screens.

File: app/src/main/java/com/kodeco/android/coordplot/country_info/CountryListData.kt [hunk 3]
```diff
@@ -9,4 +9,6 @@ object CountryListData {
     fun setData(newData: List<Country>) {
         data = newData
     }
-}
\ No newline at end of file
+
+    var dataNeedsRefreshing = false
```
Comment (zlippard) [inline app/src/main/java/com/kodeco/android/coordplot/country_info/CountryListData.kt:13]: Clever use here! This works well for this week :] As we move in to using viewmodels this week it'll start to become more apparent how you can use those classes to control these sort of stateful variables.

File: app/src/main/java/com/kodeco/android/coordplot/country_info/Flows.kt [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    val backFlow = _backFlow.asStateFlow()
+    fun tapBack() {
+        _backFlow.value += 1
+    }
+
+    private val _counterFlow = MutableStateFlow(0)
+    val counterFlow = _counterFlow.asStateFlow()
+
+    private val _currentTheme = MutableStateFlow(Theme.LIGHT)
+    val currentTheme: StateFlow<Theme> = _currentTheme.asStateFlow()
```
Comment (zlippard) [inline app/src/main/java/com/kodeco/android/coordplot/country_info/Flows.kt:39]: nit: But typically you want to put all your variables towards the top of the class, all public functions go next, then private functions go at the bottom.

File: app/src/main/java/com/kodeco/android/coordplot/country_info/components/CountersTopBar.kt [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                    },
+                    contentColor = White
+                )
+            ) {
+                Text(
+                    text = when (currentTheme.value) {
+                        Theme.LIGHT -> stringResource(R.string.go_dark)
+                        Theme.DARK -> stringResource(R.string.go_light)
+                    }
+                )
```
Comment (zlippard) [inline app/src/main/java/com/kodeco/android/coordplot/country_info/components/CountersTopBar.kt:71]: Nicely done! Great job taking this above and beyond with an additional feature to change between light/dark modes!

File: app/src/main/java/com/kodeco/android/coordplot/country_info/screens/CountryDetailsScreen.kt [hunk 6]
```diff
@@ -19,46 +19,58 @@ import androidx.compose.ui.tooling.preview.Preview
...
+                TopAppBar(
+                    title = { Text(text = countryData[countryIndex].name.common) },
+                    navigationIcon = {
+                        IconButton(onClick = {
+                            tapBack()
+                            onBackClicked()
+                        }) {
+                            Icon(
+                                imageVector = Icons.Filled.ArrowBack,
+                                contentDescription = stringResource(R.string.back_button)
```
Comment (zlippard) [inline app/src/main/java/com/kodeco/android/coordplot/country_info/screens/CountryDetailsScreen.kt:49]: Great job utilizing string resources!

File: app/src/main/java/com/kodeco/android/coordplot/country_info/screens/CountryDetailsScreen.kt [hunk 7]
```diff
@@ -19,46 +19,58 @@ import androidx.compose.ui.tooling.preview.Preview
...
+                Column(
+                    modifier = Modifier
+                        .fillMaxSize()
+                        .padding(paddingValues),
+                    horizontalAlignment = Alignment.CenterHorizontally,
+                    verticalArrangement = Arrangement.spacedBy(8.dp)
+                ) {
+                    DetailItem("Capital", countryData[countryIndex].capital?.get(0).toString())
+                    DetailItem("Population", countryData[countryIndex].population.toString())
+                    DetailItem("Area", countryData[countryIndex].area.toString())
```
Comment (zlippard) [inline app/src/main/java/com/kodeco/android/coordplot/country_info/screens/CountryDetailsScreen.kt]: nit: You should also use string reousrces for the label parameter.

**Review by zlippard — APPROVED** [review]
Overall, very well done this week @andrii-zubenko! This looks good to me, just a had a few minor nits but nothing that's blocking this MR.  Way to go above and beyond this week :]

**Comment by andrii-zubenko** [comment]
@zlippard thanks for the feedback!

**Pull request merged**